File: Advanced/Java/problem_13.java
import java.util.Stack;
 
class Test
{

    static int findMaxLen(String str)
    {
        int n = str.length();
 

        Stack<Integer> stk = new Stack<>();
        stk.push(-1);
 
        int result = 0;
 

        for (int i = 0; i < n; i++)
        {

            if (str.charAt(i) == '(')
                stk.push(i);
 
            else
            {
                if(!stk.empty())
                    stk.pop();
 
              
                if (!stk.empty())
                    result
                        = Math.max(result,
                                   i - stk.peek());
 
          
                else
                    stk.push(i);
            }
        }
 
        return result;
    }

    public static void main(String[] args)
    {
        String str = "((()()";
       

        System.out.println(findMaxLen(str));
 
        str = "()(()))))";
    
        System.out.println(findMaxLen(str));
    }
}

File: Advanced/Python/problem_13.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        cur=0
        stack=[]
        stack.append(-1)
        for i in range (len(s)):
            if s[i]=='(':
                stack.append(i)
            else:
                stack.pop()
                if len(stack)==0:
                    stack.append(i)
                else:
                    cur=max(cur,i-stack[-1])
        return cur        
